internal/upload/upload.go

```diff
@@ -0,0 +1,242 @@
+package upload
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"log/slog"
+	"maps"
+	"mime/multipart"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/buildkite/test-engine-client/internal/env"
+	"github.com/buildkite/test-engine-client/internal/version"
+	"github.com/google/uuid"
+)
+
+type runEnvMap map[string]string
+
+// Config is upload-specific configuration, but may also contain configuration
+// that is redundant with config.Config, since package upload isn't really
+// unified/integrated with the rest of bktec yet.
+type Config struct {
+	// UploadUrl is the Test Engine upload API endpoint e.g. https://analytics-api.buildkite.com/v1/uploads
+	UploadUrl string
+
+	// SuiteToken is the Test Engine upload API suite authentication token
+	SuiteToken string
+}
+
+func ConfigFromEnv(env env.Env) (Config, error) {
+	url := env.Get("BUILDKITE_TEST_ENGINE_UPLOAD_URL")
+	if url == "" {
+		url = "https://analytics-api.buildkite.com/v1/uploads"
+	}
+
+	token := env.Get("BUILDKITE_ANALYTICS_TOKEN")
+	if token == "" {
+		return Config{}, fmt.Errorf("BUILDKITE_ANALYTICS_TOKEN missing")
+	}
+
+	return Config{
+		UploadUrl:  url,
+		SuiteToken: token,
+	}, nil
+}
+
+// UploadCLI is a CLI entrypoint for uploading results to Test Engine.
+func UploadCLI(flag *flag.FlagSet, env env.Env) error {
+	cfg, err := ConfigFromEnv(env)
+	if err != nil {
+		return fmt.Errorf("configuration error: %w", err)
+	}
+
+	filename := flag.Arg(1)
+	if filename == "" {
+		return fmt.Errorf("expected filename for JUnit XML or JSON file")
+	}
+
+	info, err := os.Stat(filename)
+	if err != nil {
+		return fmt.Errorf("file does not exist: %s", filename)
+	} else if !info.Mode().IsRegular() {
+		return fmt.Errorf("not a regular file: %s", filename)
+	}
+
+	var format string
+	switch filepath.Ext(filename) {
+	case ".xml":
+		format = "junit"
+	case ".json":
+		format = "json"
+	default:
+		return fmt.Errorf("could not infer format (JUnit / JSON) from filename")
+	}
+
+	runEnv, err := buildRunEnv(env)
+	if err != nil {
+		return fmt.Errorf("unable to derive runEnv: %w", err)
+	}
+
+	slog.Info("Uploading", "key", runEnv["key"], "format", format, "filename", filename)
+
+	respData, err := Upload(cfg, runEnv, format, filename)
+	if err != nil {
+		return err
+	}
+
+	var logArgs []interface{}
+	for k, v := range respData {
+		logArgs = append(logArgs, k, v)
+	}
+	slog.Info("Upload successful", logArgs...)
+
+	return nil
+}
+
+// Upload sends test result data to Test Engine.
+func Upload(cfg Config, runEnv runEnvMap, format string, filename string) (map[string]string, error) {
+	body, err := buildUploadData(runEnv, format, filename)
+	if err != nil {
+		return nil, fmt.Errorf("Upload: %w", err)
```

Comment:
```suggestion
		return nil, err
```
We print `upload: %v` in main function already, so I think this is redundant.

Reply:
Fixed!

Before:

    2025/03/20 12:00:46 INFO Uploading key=0195b0a3-49bf-7812-af95-ff7ef0e6ff4d format=junit filename=test.xml
    Buildkite Test Engine Client: upload: Upload: opening test.xml for reading: open test.xml: permission denied

After:

    2025/03/20 12:00:46 INFO Uploading key=0195b0a3-49bf-7812-af95-ff7ef0e6ff4d format=junit filename=test.xml
    Buildkite Test Engine Client: upload: preparing upload data: opening test.xml for reading: open test.xml: permission denied